src/dune_rules/action_unexpanded.ml

```diff
@@ -89,63 +97,61 @@ end = struct
     ; dir : Path.Build.t
     }
 
-  type 'a t = env -> collector -> 'a Action_builder.t * collector
+  type 'a t = env -> collector -> ('a Action_builder.t * collector) Memo.Build.t
 
-  let return x _env acc = (Action_builder.return x, acc)
+  let return x _env acc = Memo.Build.return (Action_builder.return x, acc)
 
   let map t ~f env acc =
-    let b, acc = t env acc in
+    let+! b, acc = t env acc in
     (Action_builder.map b ~f, acc)
 
   let both a b env acc =
-    let a, acc = a env acc in
-    let b, acc = b env acc in
-    (Action_builder.both a b, acc)
+    let*! a, acc = a env acc in
+    let*! b, acc = b env acc in
```

Comment:
This is not too important but I guess this could have been `let+!` and then the subsequent `return` could be dropped.

Reply:
I guess. I used to often use `let+` when I could, but then it makes refactoring more difficult, so now I tend to always use `let*`.

Reply:
Using `let*` systematically during Fiber-related refactoring sounds fine.

I guess I'm just generally nervous about using `bind`s where `map`s are sufficient because that often comes with a cost, but I'm not sure there is any actual cost here.